### ISSUE TITLE
v1.3 I think if a picture fails to fit, it is shifted to a new location but displays as the NEXT picture!

### ISSUE THREAD

**Issue by davidc86** [sillsdev/ptx2pdf#270]
v1.3 I'm not sure exactly what is going on, but earlier I saved a copy of a configuration for testing. Then I tweaked the size and anchor for the first two pictures in GEN. Then the PDF started printing the 2nd picture twice, and not the first one! So I did a file compare of the original configuration that I was tweaking, and  the copy which I was saving. 

![image](https://user-images.githubusercontent.com/65363206/97064294-d7729d80-1572-11eb-878e-ead436f550d4.png)

One the left is the tweaked configuration, on the right is the saved copy.
Note that comparing the first line, we see that the left one now is nearly identical to the 2nd line on the left, but the right ones are different. PTXprint even copied the picture caption and the filename from the second illustration onto the first one's values. 

Yeah, it's weird. 

Anyway, that explains why the PDF had two copies of the 2nd picture -- because PTXprint had copied that entry onto the first one.

NOTE: I have been seeing something weird like since in v1.2, but so many other things were changing and misbehaving I couldn't be sure what was happening. Now that v1.3 is behaving a bit more robustly, I'm able to tweak picture settings and now I'm seeing this more clearly.

**Comment by davidc86** [comment]
I did more testing by trying to Fix the messed up entry. (But I couldn''t fix the filename, see #271 ) So I ended up making a new entry and deleting the bad one. (Not ideal.) So I printed the results and it was back to the original PDF results (which are not yet what I want, so I wanted to tweak it some more). 

But when I tweaked the anchor for just the first entry, I got two entries in the piclist: 
![image](https://user-images.githubusercontent.com/65363206/97064679-b2335e80-1575-11eb-89f0-0a9e76401c64.png)
 
I think PTXprint needs a new way to pull in the entire list and write out the entire modified list, so duplicates don't happen. The Anchor Ref. is the most modified spec in the list. Scale is the next most modified feature. (These two allow us to finely tweak the picture placement and size to get the best text flow.) I say that because it looks like maybe the Anchor Ref is used to track the picture entries (and change it creates duplicates), but that Anchor value will be changing the most. If you want to use something consistent to track entries it would be the filename. But just using numbers would work, as your software doesn't support reordering, though maybe if we swapped anchors that would switch the entries' order. 

Anyway, this is a bug.

**Comment by mhosken** [comment]
Since you are the one most interested in us supporting multiple pictures in one verse, how would you suggest that ptxprint identifies duplicates?

**Comment by davidc86** [comment]
This bug is NOT related to two pictures on one verse. This issue is a bug which takes an existing PicList file and when I edit one of the anchor point for a picture to an adjacent verse (Like changing GEN 1:25 to GEN 1:26, PTXprint DUPLICATED the entry and created another identical entry with the new Anchor Ref. (Or maybe it added a new entry and forgot to delete the old entry. I don't know how it is Saving the list is being handled, but when it saved the list out to the Hard Drive is where the error occurred. Because at the time what was showing in the display was correct. It was only after seeing a messed up print out did I track down the duplicated picture entry in the new version of the PicList file.)

**Comment by davidc86** [comment]
But to answer the 2 pictures on one verse, I would assume that anybody who wanted to do that, would basically assume each picture takes up half of the span L and R (in single column layout), maybe it'd be the same in 2-column. So if the pictures are large, they'd have to be cut in to 50% or less the full width of the span, and then one would be Span TL and the other Span TR.  (This is just an example.)

**Comment by mhosken** [comment]
It is an issue of multiple pictures in one verse, because when one changes an entry, we have to decide whether the user meant that they were making a second picture in a verse or were really wanting one of the entries on that verse to be replaced with their edits. We do a pretty good job, but it's easy to get wrong. If we only allowed one picture per verse, as implied by the anchor, then there would be no problem and we could simply key off the anchor. At the moment we have to do quite a bit of guessing and that introduces bugs. You say: I just moved a picture from one verse to another, what's so hard about that? And I agree that in this case the 'obvious' thing was clearly not obvious to the code. Aaah the joys of psychic software :-)

As to assuming that people will want a particular layout of multiple pictures in a verse; I don't think we can make such an assumption or bake that into ptxprint or rely on people doing the right thing in that regard.

**Comment by davidc86** [comment]
Well, if the two pictures on one verse is causing a problem here, then it must not be allowed. But since Paratext does not forbid it, it is likely that it will happen (especially since PT does not differentiate between "anchor" and "reference" which are two completely different animals. So while I might need two pictures REFERENCING ONE verse, I don't need them attached to that same verse. 

So PTXprint might should handle such a mess, but making the first picture is reads from the SFM file anchored to the reference verse, but the next picture for that verse gets its anchor bumped to the next verse (though it's caption reference stays the same). 

This way PTXprint would never discard a picture simply because Paratext allows users to point pictures to the same verse. And it would keep them pretty close together. 

And by eliminating the 2 pictures at one anchor problem, then the PicList could disallow it, and that (hopefully) eliminates the guess work for the code as users edit the anchors. (It would disallow a user to make a current picture's anchor identical to another picture's anchor. If the user needs to shift the one picture to the other location, then they need to move the other picture first to free up that anchor verse, then then could move the first picture to the second pictures anchor.)
(Assuming I'm understanding the problem correctly!)

**Comment by mhosken** [comment]
Moving this discussion on multiple pictures on one verse to #232. Further discussion in this issue should be about why the wrong picture is showing up in the output.

On that, could you send DG (via MP if necessary) an archive of the project perhaps with just the one book (to reduce pictures and search space).

**Comment by davidg-sil** [comment]
I've seen cases where a pictire is not included if there's no space for it. That is one reason why the debugging info moans if a picture is missing at the end of the job.  I'd be most surprised if the picture 'becomes another one'

**Comment by davidc86** [comment]
I have just submitted an Issue of duplicated picture entries #296 showing up in the PicList file when the Anchor value is changed to an earlier verse. The extra pictures show up in the exported PDF file, but their extra entry lines do NOT show up in the PTXprint TABLE while I'm still using the app). But if I close the app, and look at the PicList file, I see these duplicated picture entries that PTXprint is generating. 

I am almost certain it is changing the Anchor values that is throwing things off. I have noticed it only when I keep changing the anchor value of a single picture and test it and then change it again. 

I did not have that problem changing the anchor value of one picture, test it, move on to another picture, etc. 

It appears it MIGHT  be related to not flipping the edit control to a new picture between each change. Or it might be changing the anchor to an earlier verse reference. 

But it is a major bug, that messes up the PDF permanently. And the only way to really fix it is to edit the PicList file outside of PTXprint. 

I suspect it is because the Editor gets confused as to whether the current picture (with the tweaked anchor) is the SAME picture in the PicList file but now with a new anchor value, or is an ADDED picture (that happens to be nearly identical) that is anchored to a different anchor value that the one in the PicList file. Confused it creates a new nearly duplicate entry. But if all the pictures were simply numbered 1,2,3... Then that would give absolute reference to the picture whose settings are being edited. Adding or removing a picture would change the subsequent picture values, And the entire tabel would need to be immediately be written out and read back in again, so the table would refresh with sequential numbers again. (There may be a better way, but using Anchor Values is not it. These values change all the time as we tweak them to reposition pictures to get them to sit nicely on the page. A very iterative process.)
